Visible scroll-top button and smooth scroll on privacy-policy page

sanketbansal opened the pull request:
<!--
(Thanks for sending a pull request! .)
-->

<!--The PR title should start with "Fix #bugnum: " (if applicable), followed by a clear one-line present-tense summary of the changes introduced in the PR. For example: "Fix #bugnum: Introduce the first version of the collection editor." -->

<!-- Add issue number here. If you do not solve the issue entirely, please change the message e.g. "Addresses #IssueNumber -->
Fix #106 
Fix #107

Link to live demo: https://pr-111-evalai.surge.sh

#### Changes proposed in this pull request:

### For issue 106:
- Added Z index value to the scroll top button.
- Reduce the z-index valu of the footer to 0.
- Implemented smooth scroll for scroll top button.
- Made the main content scroll smoothly on clicking nav content.
<!-- Changes: Add here what changes were made in this issue and if possible provide links. -->

<!-- Demo Link: Add here the link where you changes can be seen. -->
<!-- Link to live demo: http://pr-XXX-evalai.surge.sh--> <!-- Replace XXX with your PR no: -->

<!-- Screenshots for the change: Add here the screenshot of the fix. ### -->
![Screenshot (40)](https://user-images.githubusercontent.com/17106489/54498313-d5267700-492b-11e9-8583-6583d6b7eaba.png)


 
### For issue 107
- Added the scroll bar in the nav container of privacy policy page.
- CSS property max-height has been added in the nav container div.

![Screenshot (44)](https://user-images.githubusercontent.com/17106489/54882104-af91f400-4e7c-11e9-8fb6-24533fbca0cc.png)

gautamjajoo commented on src/app/components/nav/footer/footer.component.scss:
```diff
@@ -11,7 +11,7 @@
   background-repeat: no-repeat;
   background-size: 100% 100%;
-  z-index: 10;
+  z-index: 0;
```
Why are you decreasing footer's `z-index`, instead increase for the button.

sanketbansal replied:
Because scroll button is child of  "<div class="privacy-content>" and <app-footer> is sibling of that div.

sanketbansal replied:
`<app-footer>`

gautamjajoo commented on src/app/components/privacy-policy/privacy-policy.component.scss:
```diff
@@ -50,10 +50,11 @@
       position: fixed;
-      top: 50%;
-      right: 4%;
+      bottom: 2%;
```
Why was this change made?

sanketbansal replied:
It positions the scroll top button at the bottom right of the privacy page

gautamjajoo replied:
Maybe the earlier one is better! You could revert this change.

gautamjajoo reviewed (commented):
Remove the changes other than solving the issue, a PR is already up for it!

sanketbansal committed:
Removed unnecessary changes

Removed build changes and privacy-policy UI changes.

RishabhJain2018 committed:
Merge branch 'master' into Privacy-policy

gautamjajoo commented:
@sanketbansal Could you add the smooth scroll as well and fix the travis build?

sanketbansal committed:
UI change

Added z-index value for scroll top button

sanketbansal committed:
Merge remote-tracking branch 'origin/Privacy-policy' into Privacy-policy

sanketbansal committed:
Added smooth scroll

implemented scrollOnTop method to add smooth scroll feature.

sanketbansal committed:
Added Linting

Linted Privacy-policy component.ts file

lunayach commented:
Hi, @sanketbansal good job on the PR. Just a couple of things;

- As you have implemented smooth scrolling, a working GIF for same could be added in PR description.
- I think you could change the PR title and it's description being more specific about what you have done.

lunayach commented on src/app/components/privacy-policy/privacy-policy.component.ts:
```diff
@@ -132,6 +132,17 @@ export class PrivacyPolicyComponent implements OnInit {
    */
   scrollToTop() {
-    this.document.body.scrollTop = this.document.documentElement.scrollTop = 0;
+    // this.document.body.scrollTop = this.document.documentElement.scrollTop = 0;
```
The commented code may be removed.

Shekharrajak replied:
Unnecessary lines should be removed before committing the changes.

lunayach commented on src/app/components/privacy-policy/privacy-policy.component.ts:
```diff
@@ -132,6 +132,17 @@ export class PrivacyPolicyComponent implements OnInit {
+    const scrollInt = setInterval(() => {
+      this.document.body.scrollTop -= bodyScrollRate;
+      this.document.documentElement.scrollTop -= eleScrollRate;
+      console.log(' Scrolling Body and Element ');
```
Log statement may also be removed.

Shekharrajak replied:
`console.log` can slow down the page. It should not be committed.

sanketbansal commented:
![evalai_edit_0](https://user-images.githubusercontent.com/17106489/54697673-f669ac00-4b53-11e9-9f5c-4499256d1bbb.gif)

sanketbansal commented:
Added the gif and changed the title of pull request.

gautamjajoo commented on CODE_OF_CONDUCT.md:
```diff
@@ -0,0 +1,44 @@
+# Contributor Code of Conduct
```
This has been added in previous commit! Please remove this

gautamjajoo commented on src/app/components/privacy-policy/privacy-policy.component.scss:
```diff
@@ -54,6 +54,7 @@
       right: 4%;
       height: 50px;
       cursor: pointer;
+      z-index: 4 !important;
```
Why was this added? Decreasing the `z-index` of footer works!

sanketbansal replied:
yeah but just to make sure that it's z-index remains high then others elements

Shekharrajak replied:
Then it should be high number like 100, right?

gautamjajoo reviewed (commented):
Everything else looks good, one minor comment.

Shekharrajak commented:
Please don't remove link https://pr-111-evalai.surge.sh/ from the PR description .

Shekharrajak commented on src/app/components/privacy-policy/privacy-policy.component.ts:
```diff
@@ -132,6 +132,17 @@ export class PrivacyPolicyComponent implements OnInit {
+      this.document.documentElement.scrollTop -= eleScrollRate;
+      console.log(' Scrolling Body and Element ');
+      if ( this.document.body.scrollTop === 0 && this.document.documentElement.scrollTop === 0 ) {
+        clearInterval(scrollInt);
```
I am not a good fan of using pure javascript in Angular application. Isn't we can do the same using 

```
scroll(el: HTMLElement) {
    el.scrollIntoView({behavior:"smooth"});
  }

```

Also, the same method can be used to scroll to down/up when user will click on contents; written in the left sidebar. WDYT ?

sanketbansal replied:
have you checked this method to solve the smooth scroll issue?

Shekharrajak commented:
I see a few other minor things (not related to this PR and issue) : 

1. Unnecessary x-scroll in the privacy-page
2. Scroll to top arrow should be disappeared when it is already on top.
3. When we click on left sidebar content, it should move up/down smoothly on the clicked section. 
Means when I click on "Third-party links" written in left side bar it should smoothly go to the "Third-party links" at the right.

WDYT?

sanketbansal committed:
Removed unnecessary changes

deleted console.logs and commented line form scrollToTop method.

gautamjajoo commented:
@Shekharrajak I was about to open an issue for 3rd point but since this PR is working on almost the same purpose, @sanketbansal Could you make the changes requested by @Shekharrajak??

sanketbansal commented:
ok, Will commit some changes for that issue.

sanketbansal committed:
Changed scrollToTop method

removed setInterval method and added scrollIntoView to scroll to top with smooth behaviour.

sanketbansal committed:
Nav content smooth scroll

Added smooth scroll for the right side content on click of nav content.

sanketbansal commented:
![cloudcv](https://user-images.githubusercontent.com/17106489/54855549-d76a4600-4d1c-11e9-8129-92275327123f.gif)

sanketbansal committed:
Fix #107

Added scroll bar for nav content.
max height property has been  added  in nav container.

sanketbansal commented:
fix #107

sanketbansal commented:
![Screenshot (44)](https://user-images.githubusercontent.com/17106489/54879564-5caa4380-4e60-11e9-9b01-009b297e7abf.png)

Shekharrajak commented:
@sanketbansal , your work is good, but few points about PRs:

- All the things that describe the PR should be in PR description.  writing fix #issue_number somewhere is PR comment is bad practice. You may edit the PR description/ GIF image with the new commit.

- commit messages are very important.  fix #issue_number, having typo error is not good in the commit message, also it should be totally relevant to your changes.  You can change the commit message and/or squash the commit as per the need.

Thanks!

gautamjajoo commented on CODE_OF_CONDUCT.md:
```diff
@@ -1,44 +0,0 @@
-# Contributor Code of Conduct
```
Why have you deleted the Code of Conduct file? Please revert this.

sanketbansal replied:
https://github.com/Cloud-CV/EvalAI-ngx/pull/111#discussion_r267420754
What were you asking to remove in that comment?

sanketbansal committed:
Added integration tests.

Number of section title elements should be equal to the number of nav elements and should be greater than zero.

Shekharrajak commented on src/app/components/privacy-policy/privacy-policy.component.spec.ts:
```diff
@@ -37,10 +41,23 @@ describe('PrivacyPolicyComponent', () => {
   });
+
+  it('should have equal number of section-title elements as nav elements', () => {
+    expect(component).toBeTruthy();
```
Already present in line 49.

sanketbansal replied:
> We can also compare the title of sidebar with title of right side (.privacy-section-title).

yeah, but the thing is not every sidebar title is same as section title.

Shekharrajak replied:
If we make it same then any issue or will it change the meaning ?

sanketbansal replied:
No, it will not change the meaning
Should i do that?

Shekharrajak replied:
Yes!

Shekharrajak commented on src/app/components/privacy-policy/privacy-policy.component.spec.ts:
```diff
@@ -37,10 +41,23 @@ describe('PrivacyPolicyComponent', () => {
+
+    expect(ALL_NAV.length).toBeGreaterThan(0);
+    expect(ALL_TARGET.length).toBeGreaterThan(0);
+    expect(ALL_TARGET.length).toBe(ALL_NAV.length);
```
This is good one.

Shekharrajak reviewed (commented):
We can also compare the title of sidebar with title of right side (.privacy-section-title).

WDYT?

sanketbansal committed:
Removed unnecessary test

component toBe(truthy) removed from the test

sanketbansal committed:
Changed section title

Made section element title same as nav element title

sanketbansal committed:
Added test for section title

Test checks the section title to be as same as nav element title or not.

sanketbansal committed:
Fixes linting issue

fixed linting issues by adding whitespace properly

Shekharrajak commented on src/app/components/privacy-policy/privacy-policy.component.spec.ts:
```diff
@@ -37,10 +41,33 @@ describe('PrivacyPolicyComponent', () => {
+
+  it('should have same section title as nav element title', () => {
+
+    const ALL_NAV = de.queryAll(By.css('.privacy-nav'));
```
Line 54, 55 is sameas line 64,65. Can you apply [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) concept here?

Shekharrajak reviewed (commented):
Other thanmy comment, it is looking good.

sanketbansal committed:
changes for privacy-policy component tests

Made ALL_NAV and ALL_TARGET variable which gets initialized before each test.

Shekharrajak commented:
Please fix travis build error, other than that it is good to go.

sanketbansal commented:
How can I see the logs of test command on travis build ?

Shekharrajak commented:
Isn't it clearly mentioned here : https://travis-ci.org/Cloud-CV/EvalAI-ngx/builds/525279140#L585 ?

sanketbansal commented:
yeah, I can see those logs but what I meant is the **test-log** file generated for **ng test command** .

Shekharrajak commented:
@sanketbansal , why not you try to understand how travis CI build works and which command it is running for testing and what changes need to be done to check test logs?

RishabhJain2018 committed:
Merge branch 'master' into Privacy-policy

RishabhJain2018 commented:
@sanketbansal Can you please look into this and try to understand the working of Travis builds. Feel free to ping @Shekharrajak @lunayach or me if you need some help.

sanketbansal commented:
yeah working on the issue

RishabhJain2018 commented:
@sanketbansal Any updates on this?

sanketbansal commented:
@RishabhJain2018  Travis build issue is resolved now.
Please review

RishabhJain2018 commented:
Thanks, @sanketbansal for fixing it quickly.
@Shekharrajak @lunayach Can you guys please review this?

lunayach reviewed (changes requested):
@sanketbansal  Currently, the surge demo is not working. Could you rebase over the master branch? #127 got merged recently, and that should make the demo work.

^ Ignore, I updated the branch using the UI below.

lunayach committed:
Merge branch 'master' into Privacy-policy

lunayach commented:
@RishabhJain2018 after #127 the demos should have been working. http://staging-evalai.cloudcv.org/api/ is the staging endpoint, Right?

sanketbansal commented:
@lunayach  All these changes had already been done. These are outdated commits.

sanketbansal committed:
Added logs in test logic

Added console logs to show title of target and nav element in test-log file

sanketbansal committed:
Added grep commands

Added grep commands in travis.yml file to show logs from test-log file

sanketbansal committed:
Changed xvfb command

Added screen size and pixel depth in xvfb command to set specific screen size of virtual display.

RishabhJain2018 commented:
There is a CORS issue with the staging server. I'll fix this in a day or two.

sanketbansal commented:
@lunayach  are there any changes needs to done?

lunayach reviewed (commented):
The code looks good. But just a small UI thing that I noticed. Shouldn't the scroll button become visible only when the user has scrolled down a bit? Currently, it just comes after some slight delay even when the user has not scrolled down.

sanketbansal committed:
Hide Scroll button on Top

Added logic to change visibility if 'privacy-policy-title'
element gets out of the view.

sanketbansal commented:
@lunayach I have added [these changes](https://github.com/Cloud-CV/EvalAI-ngx/pull/111#pullrequestreview-238719026). please review.

lunayach commented:
@RishabhJain2018 @Shekharrajak Let's get this merged!

Shekharrajak commented on .travis.yml:
```diff
@@ -14,14 +14,17 @@ before_install:
   - export DISPLAY=:99.0
   - export DEPLOY_DOMAIN=https://pr-${TRAVIS_PULL_REQUEST}-evalai.surge.sh
-  - sh -e /etc/init.d/xvfb start
+  - /sbin/start-stop-daemon --start --quiet --pidfile /tmp/custom_xvfb_99.pid --make-pidfile --background --exec /usr/bin/Xvfb -- :99 -ac -screen 0 1280x1024x16
```
I am not sure about this change. @sanketbansal can you please let us know what this line will do?

sanketbansal replied:
basically, on privacy policy page we do not show **nav bar** for **small screen size** due to which nav bar is not getting **rendered** in headless chrome environment causing values coming to be **null**. So used this line to change screen size.

Shekharrajak reviewed (approved):
Good work @sanketbansal ! Let's wait for https://github.com/Cloud-CV/EvalAI-ngx/pull/125 then we will merge this.

Shekharrajak commented:
Can we merge this PR now ? @sanketbansal

sanketbansal commented:
Yeah, It's working fine I think.

Shekharrajak committed:
Merge branch 'master' into Privacy-policy

Shekharrajak commented:
Okay! Merging it.
